### ISSUE TITLE
[discussion] Plan for the transition to GitHub

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#1380]
This is an issue to document the steps to be taken, any possible risks and mitigations, responsibilities, etc., for the transition from GitLab to GitHub. Feel free to update the list with other items.

Action list:

- [x] Discuss the idea
  - [x] done in a few meetings
- [x] Discuss the location of the new project
  - [x] initially autosubmit/autosubmit, but it's taken, so we will use [bsc-es/autosubmit](https://github.com/bsc-es/)
  - [x] Francesco B. added Bruno to the team, and we are discussing about an Autosubmit team, managed by us so that it's easier to update the list of people with write access to the repo
  - [x] Create the Autosubmit team, and add others
- [ ] Move from GitLab issue & merge request templates, to GitHub issue & pull request templates
- [ ] Move from GitLab workers CICD to GitHub Actions
  - [ ] The coverage that's handled by GitLab will have to be moved somewhere like Codecov, for instance
  - [ ] Revise "skipped" tests ( per ex, https://earth.bsc.es/gitlab/es/autosubmit/-/merge_requests/508#note_318531 )
- [ ] Choose a uniform main branch name (AS is `master`, others are `main`)
- [ ] Migrate code, issues (? or leave issues where they are) and merge requests
  - [x] autosubmit-gui
  - [x] autosubmit-api
  - [x] autosubmitconfigparser
  - [ ] autosubmit
  - This SO answer for “Transfer git repositories from GitLab to GitHub - can we, how to and pitfalls (if any)?” https://stackoverflow.com/a/22266000
  - This tool (from SO answer above) that migrates merge requests - https://github.com/piceaTech/node-gitlab-2-github?tab=readme-ov-file#transfermergerequests
- [ ] Update documentation
  - [ ] README
  - [ ] Sphinx
  - [ ] ES Wiki
  - [ ] [BSC page](https://www.bsc.es/research-and-development/software-and-apps/software-list/autosubmit) (it needs an update too)
  - [ ] References in Eduuni
  - [ ] GitLab Wiki (besides the README, there are many links there)
- [ ] GitLab Wiki (can we move part or most of the information to GitHub / README / RTD)?
- [ ] Update ReadTheDocs
- [x] Add the CITATION cff file (a comment from @manuel-g-castro in another issue reminded about this feature in GH) -- done in https://earth.bsc.es/gitlab/es/autosubmit/-/merge_requests/472
- [ ] Write a migration plan (dates for the real migrate and the switch)
- [ ] Write a communication to share with the users. We have to explain why this is a move that will benefit them in the long-term (sustainability). We have to provide the schedule of the migration, letting some time for them to ask us doubts before it's done.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 5, 2024, 13:39

Hello @kinow 

Thanks for open this issue 

> Choose a uniform main branch name (AS is `master`, others are `main`)

I would pick Main, as Master has been deprecated for a while. 

> Migrate issues (? or leave issues where they are)

I would migrate the issues if possible. There are a lot, but can we migrate only non-completed issues? 

> Migrate merge requests (? not sure if this is supported)

If it is not possible, I guess we can always do a git clone and git push to a new branch in github. But I'm not sure if GitHub will understand it when we try to merge it in GitHub. We would have to do a test. I think we should have at least all of them rebased before migrate

An alternative would be to have all merge requests merged into the master branch... but some branches like yours with pyflows would be hard to do before the migration. Btw, when do we have to do the migration? Any estimated date? @mcastril 

> Create the Autosubmit team, and add others

+1

**Comment by kinow** [comment]
In GitLab by @LuiggiTenorioK on Aug 16, 2024, 14:35

Thanks @kinow!

The plan seems pretty nice. However, I think the main concern will be related to the issues/merge requests.

Are we going to have both issues boards open? I think it could be chaotic to have both boards open because we are going to have conflicts with the issues/mr numbers especially when linking them to the commits messages.

One possible approach is that once we migrate issues to GitHub, **we close the issue board of GitLab**. This makes sense as all the users from GitLab will be able to create issues on GitHub but not in the opposite direction. I have other approaches in mind but we can discuss them later.

Other than that, we can add to the list: _"Add a Contributing Guideline"_. I found a really nice example [here](https://github.com/jessesquires/.github/blob/main/CONTRIBUTING.md). This will encourage and help external devs to contribute with Autosubmit which is a perk we will have by doing this transition. For this, maybe we can start by defining a coding style (maybe guided by a formatter like Ruff).

**Comment by kinow** [comment]
mentioned in commit 20e0c2e12b849fa770c411e2af6a1908189a3662

**Comment by kinow** [comment]
mentioned in commit 9ee6ca5c6976a5f0e12b4a43aa1261efefe53122

**Comment by kinow** [comment]
mentioned in commit 0c07324d37f50a4d6171da036df4d32c342947e3

**Comment by kinow** [comment]
Created `autosubmit` team. Created first repository for Autosubmit GUI, https://github.com/BSC-ES/autosubmit-gui.

**Comment by kinow** [comment]
Added `autosubmit` team to `autosubmit-gui` as "maintainer". The only role above that is "admin". The main difference is that admin has destructive/sensitive power (can delete everything). We can add admins manually there later. Francesco is also a org admin and can add/remove others.

**Comment by kinow** [comment]
Followed the instructions to mirror the code. Takes 1 minute :slight_smile: 

https://github.com/piceaTech/node-gitlab-2-github

Now missing the merge requests, issues, etc.

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/b4c70bf266af3c77e96f71395e66801b/image.png)

**Comment by kinow** [comment]
In GitLab by @edgano on Aug 28, 2024, 11:15

yes, I found the same repo to migrate issues, PR, etc...
Nice to hear that it worked!

**Comment by kinow** [comment]
In GitLab by @LuiggiTenorioK on Aug 28, 2024, 11:38

Once the migration is finished, should we archive (lock) the repository in GitLab? WDYT?

> Archiving the project makes it entirely read-only. It is hidden from the dashboard and doesn't display in searches. The repository cannot be committed to, and no issues, comments, or other entities can be created. [Learn more](https://earth.bsc.es/gitlab/help/user/project/settings/index#archive-a-project).

**Comment by kinow** [comment]
Yup, +1 to lock, once we are happy with the migration.

@LuiggiTenorioK maybe we can do that now, or maybe tomorrow after the AS meeting. To give a chance for @dbeltrankyl , @mcastril , others to read this issue and have a look at what the issues/MR's/etc look like after the migration

**Comment by kinow** [comment]
In GitLab by @LuiggiTenorioK on Aug 28, 2024, 11:44

I think I'll do it now to prevent new activity. We can unarchive it anytime if it doesn't seem to be the right way.

**Comment by kinow** [comment]
Migration for autosubmit-gui is done, https://github.com/BSC-ES/autosubmit-gui

**Migration notes**:

My settings, minus tokens:

```typescript
import Settings from './src/settings';

export default {
  gitlab: {
    url: 'https://earth.bsc.es/gitlab/',
    token: '111',
    projectId: 618,
    listArchivedProjects: false,
    sessionCookie: "222",
  },
  github: {
    // baseUrl: 'https://github.mycompany.com:123/etc',
    // apiUrl: 'https://api.github.mycompany.com',
    owner: 'BSC-ES',
    ownerIsOrg: true,
    token: '333',
    token_owner: 'kinow',
    repo: 'autosubmit-gui',
    recreateRepo: false,
  },
  s3: {
    accessKeyId: null,
    secretAccessKey: null,
    bucket: null,
    region: null,
  },
  usermap: {},
  projectmap: {},
  conversion: {
    useLowerCaseLabels: true,
  },
  transfer: {
    description: true,
    milestones: true,
    labels: true,
    issues: true,
    mergeRequests: true,
    releases: true,
  },
  dryRun: false,
  useIssueImportAPI: true,
  usePlaceholderMilestonesForMissingMilestones: true,
  usePlaceholderIssuesForMissingIssues: true,
  useReplacementIssuesForCreationFails: true,
  useIssuesForAllMergeRequests: false,
  filterByLabel: undefined,
  trimOversizedLabelDescriptions: true,
  skipMergeRequestStates: [],
  skipMatchingComments: [],
  mergeRequests: {
    logFile: './merge-requests.json',
    log: false,
  },
} as Settings;

```

I executed a few of times with `dryRun: true`, adjusting it, and looking at the errors. Had a couple of errors that I had to look at the source code to figure out what was missing (token, url, project id, etc.).

The migration took about 10 minutes for Autosubmit GUI (next time will use `time`).

@LuiggiTenorioK pointed out an interesting thing, all issues are created with my user, as I'm the one running the code and calling the GH API. But the comments always start with something like "mcastril on GitLab etc etc"

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/f5a826f5345eb6bb2461175fbd65cef1/image.png)

**Comment by kinow** [comment]
>I think I'll do it now to prevent new activity. We can unarchive it anytime if it doesn't seem to be the right way.

Sounds good, it's not like anything will break/stop working by doing that :slight_smile: thank you Luiggi!

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Aug 28, 2024, 11:50

we should probably tell users that they will have some unexpected issues due to the migration, no?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 28, 2024, 11:51

Thanks Bruno for the migration, 

> @LuiggiTenorioK pointed out an interesting thing, all issues are created with my user, as I'm the one running the code and calling the GH API. But the comments always start with something like "mcastril on GitLab etc etc"

I don't see an issue with that then 

I have one question, 

What happens with the references between GUI and other Gitlab projects? ( like Autosubmit ).  Will the links still work?

**Comment by kinow** [comment]
>What happens with the references between GUI and other Gitlab projects? ( like Autosubmit ). Will the links still work?

Excellent question. We will have to find some issues with those links to check. Hadn't thought about that, @dbeltrankyl !

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 28, 2024, 11:52

> we should probably tell users that they will have some unexpected issues due to the migration, no?

@manuel-g-castro Yes, I think that too. Also, we should announce where they should open new issues, follow discussions... etc

**Comment by kinow** [comment]
>we should probably tell users that they will have some unexpected issues due to the migration, no?

@manuel-g-castro yup, just commented on Slack:

> [ ] inform users (over, and over, and over, until everybody is aware)

We will have to tell users on MWT meeting... CES meeting... ES department meeting... DestinE channels... and then tell again when they complain they were not aware :rofl:

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Aug 28, 2024, 11:55

> We will have to tell users on MWT meeting... CES meeting... ES department meeting... DestinE channels... and then tell again when they complain they were not aware :rofl:

I think we should also send a mail to the bsc-es as it is a pretty major change, and maybe also migrate the API before doing anything as they're more or less coupled, no?

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Aug 28, 2024, 11:56

But like, we should tell them now. Since they will no be able to open issues/comment etc due to the locking.

**Comment by kinow** [comment]
Yup, we locked now just to be sure nothing is added, but we will discuss tomorrow morning what's the best course of action.

In reality, we could decide to re-do the migration if others prefer to have issues some other way, import things manually, avoid importing something, etc.... so this migration is not exactly done.

We will have ~24 hours of blackout in the autosubmit-gui repo just until we clear that up in the AS meeting... but I think that should be fine, as the repository has very low activity (exactly why we are choosing this repo, and because the issues/mrs/code/etc is simpler to migrate).

**Comment by kinow** [comment]
marked the checklist item **Discuss the location of the new project** as completed

**Comment by kinow** [comment]
marked the checklist item **Create the Autosubmit team, and add others** as completed

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 28, 2024, 13:30

> I would migrate the issues if possible. There are a lot, but can we migrate only non-completed issues?

> One possible approach is that once we migrate issues to GitHub, **we close the issue board of GitLab**. This makes sense as all the users from GitLab will be able to create issues on GitHub but not in the opposite direction. I have other approaches in mind but we can discuss them later.

Sorry for not having reacted to this before. We have some important issues on our hands.

I don't have any inconvenience about making your lives easier, but a decision like this has to be taken with all the pros/cons in consideration. We have to look at what's more convenient for us and for users.

In our earliest discussions, I think we saw it reasonable to start with two boards (GitHub and GitLab) and see how manageable it is. I think that we said that we didn't think that there was going to be an avalanche of issues in GitHub from one day to the next and that it could be manageable. I also understand that with the projects we are involved in now, we can get more issues than expected, and hopefully more, from users that we don't know yet. I also think that @LuiggiTenorioK point about the duplicity is sensible.

However, we also have to take into account that the BSC users don't necessarily have a user in GitHub and that GitHub also enforces 2FA, for which we should train users if we make the move. Also, I think that it won't be so easy to link issues from other projects and that if we close the board, we should see if the closed issues remain there. Otherwise, we'll lose the links in all the issues in the Auto-Model projects that point to issues in Autosubmit. The same could happen with the mentions in commits and merges if the hashes were not preserved. These are just some ideas that come to my mind in 5 minutes, but I am sure there will be more, and we have to take all into consideration to make the decision. 

There should be solutions for all this, but it has to be considered. Could you make a table with pros and cons of the two options? Having one only board in GitHub, and having two boards. For the second option, consider too the possibility of a transition phase, but we could discard it too if it was unmanageable.

CC @f-macchia

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 28, 2024, 13:54

> As I mentioned above, we just have to unarchive (doing right now) to keep it as it was.

I think I had read it, but I wanted to be sure. There hasn't been activity on this issue in the last two weeks, and in 3 hours, everything unrolled quite fast, and I was responding to other issues in the meantime.

> Remember that what we have in GitHub is a clone and archiving the repo is just a way to prevent duplicated efforts and maintain consistency in our issues/PR lists.

I understand the reasons, as I have just said in my [comment above](https://earth.bsc.es/gitlab/es/autosubmit/-/issues/1380#note_302622). But this is not a private project. It's software used in production, including operational forecasts, and users (our client) are a critical part of the equation, as important as the developers.

I didn't say not to close the repositories, but I said to do it after all the aspects are considered and the users are informed. You have colleagues in the team (Eric, Gilbert, Alejandro, Aina, Leo...) who are assisting all the users, and I don't know how informed they are about the migration being done today.

**Comment by kinow** [comment]
At least now we know that technically we are able to migrate everything, and we know what it looks like. The repository is not archived any longer, so we can discuss the other points over the next days, until we are happy, and then we should be able to migrate GUI + API on the same day, and announce it.

Autosubmit is a bit larger, and I would do that another day or week (each migration consumes a lot of my user rate limits... just to avoid being blocked).

**Comment by kinow** [comment]
In GitLab by @LuiggiTenorioK on Aug 28, 2024, 15:04

> However, we also have to take into account that the BSC users don't necessarily have a user in GitHub and that GitHub also enforces 2FA, for which we should train users if we make the move.

Creating a user in GitHub doesn't seem to be a problem. Also, 2FA is only enforced by certain conditions (I personally don't have it enabled) which are not imposed for normal users.

> Also, I think that it won't be so easy to link issues from other projects and that if we close the board, we should see if the closed issues remain there. Otherwise, we'll lose the links in all the issues in the Auto-Model projects that point to issues in Autosubmit. The same could happen with the mentions in commits and merges if the hashes were not preserved.

Yes, there is a misunderstanding there. We wanted to archive the repo (make it read-only) instead of closing the issue board to avoid this.

> There should be solutions for all this, but it has to be considered. Could you make a table with pros and cons of the two options? Having one only board in GitHub, and having two boards. For the second option, consider too the possibility of a transition phase, but we could discard it too if it was unmanageable.

Here are the options we have:

(I'm omitting the pros of migrating to GitHub (which are a lot) since they are implicit in all the options)

## Option 1: Closing GitLab issues/PR boards

### Pros:

* Completely ensures that we only have these issues in one place (GitHub)
* No branching (e.g. issue "#XX" in GitLab could be different from issue "#XX" in GitHub) of the issue list 
* Code updates will be available in both repos

### Cons:

* Issues can't be accessed in GitLab. So, users will have to go to GitHub to read them.
* No retrocompatibility
* Users have to log in to GitHub to interact with the issues.

**(This option was discarded since we want to still have access to keep the reference in other repositories)**

## Option 2: Archiving GitLab repository

### Pros:

* No branching of the issue list
* Issues can still be referenced by other repositories in GitLab
* Code can be cloned. So, old guides, user scripts, and integrations (retrocompatibility) will work.

### Cons:

* Newer issues and code updates (no forward compatibility) will not be available on GitLab. Making it the "legacy" repo.
* Users have to log in to GitHub to interact with the issues.


## Option 3: Keeping both issues list

### Pros:

* Issues can still be referenced by other repositories in GitLab
* GitLab users can still interact with the repository

### Cons:

* Branching (e.g. issue "#XX" in GitLab could be different from issue "#XX" in GitHub) of the issue list which will make it harder to track them and reference them in the code.
* Will make the transition harder since comments/issues/PRs that are on GitLab but not in GitHub and vice-versa are not easy to migrate (or even track them).
* Will make the Merge Requests process harder. Since we have to merge in one repo and close it in another without merging. It will be messy to review them too.


Personally, I agree to go with **Option 2** since Option 3 will be unmanageable at some point. Also, with option 2 we will ensure retro compatibility (but not forward compatibility) to the integrations we have right now. However, I completely agree that **this should be widely communicated to everyone who is using Autosubmit nowadays to make it work**.

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 28, 2024, 16:13

Thanks Luiggi for collecting the list

> Creating a user in GitHub doesn't seem to be a problem. Also, 2FA is only enforced by certain conditions (I personally don't have it enabled) which are not imposed for normal users.

Good to know that 2FA is not enforced. I had enabled long time ago after receiving multiple notifications for doing so.

Even if it's not a problem, it's an extra step, and we have to be able to explain to the users why this is a move that will benefit them, besides the annoyance of handling an extra account.

> Yes, there is a misunderstanding there. We wanted to archive the repo (make it read-only) instead of closing the issue board to avoid this.

That makes sense @LuiggiTenorioK. I haven't closed very few repositories, and I knew that closing the issue board removes access to them. I haven't used Archive before, and it looks like the best option for this case. 

I also think Option 2 is the most sensible.

> However, I completely agree that **this should be widely communicated to everyone who is using Autosubmit nowadays to make it work**.

Exactly. We should let everybody know in advance and also give them some time to ask us questions before we do it. There can be some cases that we are not considering, although having all the information available in read-only mode will facilitate things

**Comment by kinow** [comment]
In GitLab by @LuiggiTenorioK on Aug 29, 2024, 12:51

From today's meeting: We have to map the GitLab to GitHub account names before doing the migration of issues/PRs.

Following that, we can proceed in this way:

1. We can start by asking for the GitHub accounts. (Maybe next week)

    - Share spreadsheet of the accounts mapping: https://docs.google.com/spreadsheets/d/1Vb2BXlQ8vAtvfNhwT71u1zhO1qj951pcYIDF-6T77LA/edit?gid=0#gid=0 
    - Send it to Slack general channel and bsc-es@ bsc.es mailing list
    - Communicate the intention to do the migration to GitHub

```
Dear Autosubmit users,

<motivation of the migration>. 

<ask and share link of the spreadsheet>.

Best regards,
Autosubmit Team
```

2. Communicate the migration schedule to the department. (Could be immediately after step 1 or at the same time)
    
    - Send it to the general Slack channel and the mailing lists: bsc-es@ bsc.es, autosubmit@ bsc.es (Maybe some more? @mcastril)
    - Comment the implications: create a GitHub user, what is going to happen to the GitLab repo (read-only, etc)
    - Comment it on the CES and MWT meetings.
    - Make a quick example to follow your migrated issues in GitHub: Should be as simple as changing its URL like: 
    
```
earth.bsc.es/gitlab/es/autosubmit-gui/-/issues/115 -> github.com/BSC-ES/autosubmit-gui/issues/115)
```

```
Dear Autosubmit users,

<motivation of the migration>. 

<Migration plan and schedule>.

<Consequences of the migration>.

<Steps to do after the migration>.

Best regards,
Autosubmit Team

```

3. Migrate GUI repo. Maybe we can do it two weeks from the communication but not wait too long so it can be done at a time when there is not much-expected activity.

    - Review the migration
    - Archive the GitLab repo
    - Communicate successful migration

4. Migrate API repo. Could be done on the same day as the GUI (WDYT? @kinow). Same steps as step 3.

5. Migrate Autosubmit repo after the API/GUI migration. 

    - Maybe some weeks after doing the API/GUI repos migration. So we can solve possible unexpected issues.
    - Same steps in steps 3 & 4.

WDYT?

**Comment by kinow** [comment]
In GitLab by @mcastril on Sep 6, 2024, 17:00

> * Send it to Slack general channel and [bsc-es@ bsc.es](mailto:bsc-es@ bsc.es) mailing list

Done today! We may ask as a reminder by Slack in the following days.

**Comment by kinow** [comment]
Inboxes are empty, no more meetings, and not enough time to review zombies or postgres today. So will spend some time on this now :slight_smile: aiming at at least having the script that lists the existing users, and then run it against the autosubmit4-config-parser repository, which might be the simplest one to be moved.

**Comment by kinow** [comment]
@LuiggiTenorioK, here's what I came up to get the list of users in MRs, issues, notes, comments, descriptions, etc.: https://github.com/piceaTech/node-gitlab-2-github/pull/205

Here's my current `settings.ts`, with `# note-to-luiggi` comments that are not valid, but just to explain things:

```ts
import Settings from './src/settings';

export default {
  gitlab: {
    url: 'https://earth.bsc.es/gitlab/',
    token: '???',  # Same token I use to commit
    projectId: 1097,  # Copied from the autosubmit4-config-parser project page, below the project name
    listArchivedProjects: false,
    sessionCookie: "????",  # Copied from _gitlab_session cookie in the browser
  },
  github: {
    // baseUrl: 'https://github.mycompany.com:123/etc',
    // apiUrl: 'https://api.github.mycompany.com',
    owner: 'BSC-ES',
    ownerIsOrg: true,
    token: '????',  # a GH issue created following their readme instructions
    token_owner: 'kinow',  # your GH user
    repo: 'autosubmit4-config-parser',  # the future repo, must exist even for dry-run
    recreateRepo: false,
  },
  s3: {
    accessKeyId: null,
    secretAccessKey: null,
    bucket: null,
    region: null,
  },
  usermap: {  # this is where we have to add all the from->to users
    'bdepaula': 'kinow',
  },
  projectmap: {},
  conversion: {
    useLowerCaseLabels: true,
  },
  transfer: {
    description: true,
    milestones: true,
    labels: true,
    issues: true,
    mergeRequests: true,
    releases: true,
  },
  dryRun: true,  # I'm using this to test things first
  exportUsers: true,  # what I added in the branch from the PR I sent you
  useIssueImportAPI: true,
  usePlaceholderMilestonesForMissingMilestones: true,
  usePlaceholderIssuesForMissingIssues: true,
  useReplacementIssuesForCreationFails: true,
  useIssuesForAllMergeRequests: false,
  filterByLabel: undefined,
  trimOversizedLabelDescriptions: true,
  skipMergeRequestStates: [],
  skipMatchingComments: [],
  mergeRequests: {
    logFile: './merge-requests.json',
    log: false,
  },
} as Settings;
```

With this file created in the root dir of the project, after `npm install`, you can run `npm run start` and wait while it fetches data, processes everything, and then writes to API if `dryRun: false`, and creates the users.txt if `exportUsers: true`. :thumbsup: 

Here are the users found running this on autosubmit4-config-parser:

- bdepaula
- rahme1
- mcastril
- dbeltran
- ltenorio
- mgimenez
- kinow
- agayayav
- froura
- mandres
- malturjman
- agarci3
- larriola

Not sure how my GitHub user ended up there (kinow), but harmless.

**Comment by kinow** [comment]
In GitLab by @LuiggiTenorioK on Sep 27, 2024, 14:25

Thanks Bruno!

I've used your changes and in the API and GUI, I got 25 and 36 users, respectively.

[users-api-20240927.txt](https://earth.bsc.es/gitlab/es/autosubmit/uploads/adf677b43e5d086965ee1a8b87722c9f/users-api-20240927.txt)

[users-gui-20240927.txt](https://earth.bsc.es/gitlab/es/autosubmit/uploads/890f1d449a6f1380edcaf74c5e3d0a24/users-gui-20240927.txt)

**Comment by kinow** [comment]
Huh, my branch got merged into that library... didn't expect anyone with merge rights to review that, especially on a Friday, and merge it within a few hours I finished it :rofl:

**Comment by kinow** [comment]
@LuiggiTenorioK , from the previous comment about config-parser, left is GitLab, right is GitHub:

* bdepaula - kinow
* rahme1 - rohanali1798
* mcastril - mcastril
* dbeltran - dbeltrankyl
* ltenorio - LuiggiTenorioK
* mgimenez - manuel-g-castro
* kinow - kinow (no idea why my GH is on GL :shrug: )
* agayayav - ainagaya
* froura - franra9
* mandres - mandresm
* malturjman - Hussam-Turjman
* agarci3 - agarci3
* larriola - Lerriola
* isimomun - isimo00

So I think we have all users here?

**Comment by kinow** [comment]
(We may have to re-run it and update, quite sure we will have at least one new user since our last run of the typescript tool -- @isimo00 :slight_smile: )

**Comment by kinow** [comment]
In GitLab by @LuiggiTenorioK on Oct 10, 2024, 11:05

That was fast! I was writing the pending list haha

> So I think we have all users here?

Yes, we do! (Just @isimo00 pending)

Now, we need:

- [x] Have Owner role on autosubmit4-config-parser (to archive it later) (@mcastril) 
- [x] Create the repo on BSC-ES GitHub and have enough permissions
- [ ] Get approved by @mcastril

**Comment by kinow** [comment]
isimomun - isimo00

**Comment by kinow** [comment]
In GitLab by @LuiggiTenorioK on Oct 10, 2024, 11:26

> kinow - kinow (no idea why my GH is on GL :shrug: )

I realized that is because you are also storing the usernames after they are mapped here:

```ts
    if (hasUsermap) {
      reString = '@' + Object.keys(settings.usermap).join('|@');
      str = str.replace(
        new RegExp(reString, 'g'),
        match => '@' + settings.usermap[match.substring(1)]
      );
    }

    // Store usernames found in the text
    const matches: Array<string> = str.match(usernameRegex);
    if (matches && matches.length > 0) {
      for (const username of matches) {
        this.users.add(username.substring(1));
      }
    }
```

I think is just a matter of moving the function before the mapping :laughing:

**Comment by kinow** [comment]
That might be needed for the migration of api/gui & autosubmit...

**Comment by kinow** [comment]
In GitLab by @LuiggiTenorioK on Nov 4, 2024, 11:45

marked the checklist item **autosubmitconfigparser** as completed

**Comment by kinow** [comment]
In GitLab by @LuiggiTenorioK on Nov 4, 2024, 13:36

Ok, I guess we can still contact some of them (I already sent a message to Valentina Sicardi). For those who we cannot reach, I'm thinking of just mapping them as `@ username` (with the space) so it doesn't bother anyone and we don't have to mess up the migration tool code :sweat_smile: 

I'll send an email later announcing the migration to be done next Tuesday (Oct 12th) with the list of the missing usernames.

**Comment by kinow** [comment]
Either `@` + `space` or no `@`. Either sound good to me.

>we don't have to mess up the migration tool code :sweat_smile:

Yeah, I think we don't need to suggest this feature to the upstream repo :sweat_smile: Even though maybe someone else could have a similar issue, this one we can just modify our script and use it for GUI/API/AS (we will have the same issue in AS for sure).

**Comment by kinow** [comment]
Added a few more. emoreno, portega, cgonzal6, and maybe ebergas might not be too hard to find whether they have a github user or not.

I think gbonet and jberlin have github users, but I guess we will have to ask them directly for that.

The other I don't know. Maybe Dani or Miguel might know them from their time at the BSC (or maybe they still work there and I am just not aware :laughing: )

**Comment by kinow** [comment]
There are a few pending things to do now, so I'm leaving this issue open just until we fix #2030, I add the 2 PR's from Edgar, and then update the remaining docs.

**Comment by dbeltrankyl** [comment]
> , I add the 2 PR's from Edgar, and then update the remaining docs.

Oh, I commented about this today. Didn't remember this comment

> I'm leaving this issue open just until we fix https://github.com/BSC-ES/autosubmit/issues/2030,

Maybe we should push this to be done soon,

**Comment by kinow** [comment]
I created a pending pull request, but there is another PR that needs to be created.

Someone also needs to go through the checklist at the top of this issue and confirm what's done, and what's pending, and then either fix or create separate or sub issues.

**Comment by dbeltrankyl** [comment]
Updating the readme #2116

**Comment by kinow** [comment]
And today I will create the other pending pull request...

**Comment by kinow** [comment]
Pending pull requests created.

**Comment by dbeltrankyl** [comment]
I'll assign this to 4.2

It will be good to have the wiki here, there were some issues with the /esarchive during this week, and I had to refer them to the old GitLab (operational errors section)

**Comment by kinow** [comment]
> [BSC page](https://www.bsc.es/research-and-development/software-and-apps/software-list/autosubmit) (it needs an update too)

Contacted @f-macchia via Slack and she's checking who can update it 🙂 (thanks!)

**Comment by kinow** [comment]
>  ES Wiki

Updated AS, API, and GUI repositories at

- https://earth.bsc.es/wiki/doku.php?id=working_groups:computational_earth_sciences
- https://earth.bsc.es/wiki/doku.php?id=main_page:starting_at_es
- https://earth.bsc.es/wiki/doku.php?id=library:git

**Comment by kinow** [comment]
>  GitLab Wiki (besides the README, there are many links there)

We have the following pages in BSC GitLab's Wiki:

- [x] Code coverage
  - IMHO, not needed as we have this in Codecov and GitHub. The old page was needed as it was difficult to configure GitLab. We wrote it down just in case others had to do that in their repositories in GitLab.
- [x] Deployment
  - Not sure if up to date, but copied it to https://github.com/BSC-ES/autosubmit/wiki/Deployment
- [x] Issues documenting different aspects
  - Lists old issues in the project, not sure if pertinent to the current version, not migrated
- [x] Leaflet
  - Page with a pdf, last updated 9 years by Manubens-Gil, not migrated
- [x] Possible Operational Problems and Solutions
  - https://github.com/BSC-ES/autosubmit/wiki/Possible-Operational-Problems-and-Solutions
- [x] Running Autosubmit in Earth Sciences
  - This looks like something for our internal Wiki
- [x] Testing_Suite
  - Last updated 4 years ago, seems more like something for our internal wiki, or maybe testing suite wiki
- [x] Updating ReadTheDocs Autosubmit documentation
  - Contains info on SPhinx, and updating our docs; not needed, I think, as we are aware of how to update our docs, and there's external info available from Sphinx/RTD docs.
- [x] Visual Identity
  - List of logos to be used in presentations and other media; superseded by our website page https://autosubmit.readthedocs.io/en/master/media/index.html
- [x] [DestinE] Autosubmit VM on Lumi
  - Information is not quite up to date, but we have that in the CSC Wiki, I think we do not need to have it duplicated here (more work for us to keep that up to date)
- [x] background
  - This appears to have been very well written, and to contain useful information, although not sure if up to date. I created an issue to see if we can use some of that in our RTD docs -- https://github.com/BSC-ES/autosubmit/issues/2566
- [x] bibtex
  - contains the bibtext for our main publication, that will be fixed in #1355 
- [x] databases
  - contains only the text "under construction", not migrated
- [x] development
  - contains some information about development, pypi, modules, it seems to be covered in the deployment + CONTRIBUTING.md, not migrated
- [x] dissemination
  - Added the link to this page in https://github.com/BSC-ES/autosubmit/issues/1440, to add publications/etc. to our website
- [x] Home
  - Not needed, we have one in GH
- [x] installation
  - Does not contain information about AS4, only AS3, and AS2. Not migrated.
- [x] migrate experiments
  - Not sure if useful, nor if we already have that in our website. I've created an issue to track that, https://github.com/BSC-ES/autosubmit/issues/2567
- [x] versions
  - Appears to contain info mixed about what's available in BSC envs, what changed in each version, and some links. The same info can be found, up to date, in our changelog, and in our website/wiki. Not migrated (one less place to update after a release)
